timeutil: avoid parsing non-numeric strings as ints/floats

rafiss opened the pull request:
### ttljob: fix sub-benchmark of BenchmarkTTLExpiration

The b.Run call defines a new subtest, and each subtest should run for
b.N iterations.

### ttljob: make BenchmarkTTLExpiration setup quicker 

### timeutil: avoid parsing non-numeric strings as ints/floats

Constructing the error of strconv.ParseInt and strconv.ParseFloat is
relatively expensive. It allocates a new byte array then clones the
input string into it. That makes it a bad thing to do eagerly -- it's
better to check if the string has non-numeric characters and avoid
paying the cost of constructing the parsing error if we can.

```
name                                  old time/op    new time/op    delta
TTLExpiration/query=tz_conv-10           203ms ± 5%     151ms ± 3%  -25.51%  (p=0.000 n=10+10)
TTLExpiration/query=interval_math-10     108ms ± 7%     106ms ± 1%     ~     (p=0.423 n=9+8)

name                                  old alloc/op   new alloc/op   delta
TTLExpiration/query=tz_conv-10           116MB ± 0%      64MB ± 0%  -45.42%  (p=0.000 n=9+10)
TTLExpiration/query=interval_math-10    33.0MB ± 0%    33.0MB ± 0%     ~     (p=0.243 n=9+10)

name                                  old allocs/op  new allocs/op  delta
TTLExpiration/query=tz_conv-10           4.00M ± 0%     1.50M ± 0%  -62.46%  (p=0.000 n=7+10)
TTLExpiration/query=interval_math-10      251k ± 0%      251k ± 0%     ~     (p=0.920 n=9+10)
```

Epic: None
Release note: None

rafiss committed:
ttljob: fix sub-benchmark of BenchmarkTTLExpiration

The b.Run call defines a new subtest, and each subtest should run for
b.N iterations.

Release note: None

rafiss committed:
ttljob: make BenchmarkTTLExpiration setup quicker

Release note: None

rafiss committed:
timeutil: avoid parsing non-numeric strings as ints/floats

Constructing the error of strconv.ParseInt and strconv.ParseFloat is
relatively expensive. It allocates a new byte array then clones the
input string into it. That makes it a bad thing to do eagerly -- it's
better to check if the string has non-numeric characters and avoid
paying the cost of constructing the parsing error if we can.

```
name                                  old time/op    new time/op    delta
TTLExpiration/query=tz_conv-10           203ms ± 5%     151ms ± 3%  -25.51%  (p=0.000 n=10+10)
TTLExpiration/query=interval_math-10     108ms ± 7%     106ms ± 1%     ~     (p=0.423 n=9+8)

name                                  old alloc/op   new alloc/op   delta
TTLExpiration/query=tz_conv-10           116MB ± 0%      64MB ± 0%  -45.42%  (p=0.000 n=9+10)
TTLExpiration/query=interval_math-10    33.0MB ± 0%    33.0MB ± 0%     ~     (p=0.243 n=9+10)

name                                  old allocs/op  new allocs/op  delta
TTLExpiration/query=tz_conv-10           4.00M ± 0%     1.50M ± 0%  -62.46%  (p=0.000 n=7+10)
TTLExpiration/query=interval_math-10      251k ± 0%      251k ± 0%     ~     (p=0.920 n=9+10)
```

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/141506)
<!-- Reviewable:end -->

spilchen reviewed (commented):
These improvements look good. And thanks for fixing the benchmark!

> Reviewed 1 of 1 files at r1, 1 of 1 files at r2, 1 of 1 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/141506#-:-OJ4TuMm0RwZvYGc5F6W:b-23k7kg)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

___
*[`pkg/util/timeutil/time_zone_util.go` line 71 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/141506#-OJ4TlFK3qRrRIlfPK2z:-OJ4TlFK3qRrRIlfPK3-:bpshkxr) ([raw file](https://github.com/cockroachdb/cockroach/blob/93449886743099e2ef2d05ebd9dcc15da7df0115/pkg/util/timeutil/time_zone_util.go#L71)):*
> ```Go
> 	// contains any non-numeric characters to see if we can skip attempting to
> 	// parse it as a timezone offset.
> 	containsNonNumeric := strings.ContainsAny(locStr, "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ/")
> ```

nit: is there a reason why we check for `/` (last character to search for)


<!-- Sent from Reviewable.io -->

rafiss reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/141506)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @spilchen)

___
*[`pkg/util/timeutil/time_zone_util.go` line 71 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/141506#-OJ4TlFK3qRrRIlfPK2z:-OJ4USjD67x6WEPW2kUS:bnxqg8k) ([raw file](https://github.com/cockroachdb/cockroach/blob/93449886743099e2ef2d05ebd9dcc15da7df0115/pkg/util/timeutil/time_zone_util.go#L71)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

nit: is there a reason why we check for `/` (last character to search for)
</blockquote></details>

my reasoning was that most timezone names include a `/`. for example: https://github.com/cockroachdb/cockroach/blob/ddd0076bf430e74ad79da369c163854146afa8ab/pkg/util/timeutil/lowercase_timezones_generated.go#L10-L22


<!-- Sent from Reviewable.io -->

rafiss committed:
timeutil: skip ContainsAny check if timezone is UTC

This provides a minor speedup compared to the parent commit.

```
name                                  old time/op    new time/op    delta
TTLExpiration/query=tz_conv-10           145ms ± 2%     144ms ± 2%  -1.04%  (p=0.024 n=9+9)
TTLExpiration/query=interval_math-10     104ms ± 2%     104ms ± 1%    ~     (p=0.968 n=9+10)

name                                  old alloc/op   new alloc/op   delta
TTLExpiration/query=tz_conv-10          63.4MB ± 0%    63.4MB ± 0%    ~     (p=0.968 n=9+10)
TTLExpiration/query=interval_math-10    33.0MB ± 0%    33.0MB ± 0%    ~     (p=0.661 n=10+9)

name                                  old allocs/op  new allocs/op  delta
TTLExpiration/query=tz_conv-10           1.50M ± 0%     1.50M ± 0%    ~     (p=0.393 n=10+10)
TTLExpiration/query=interval_math-10      251k ± 0%      251k ± 0%    ~     (p=0.927 n=10+10)
```

Release note: None

spilchen reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/141506#-:-OJ4bSHR6ArPYjHZ7c3h:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @rafiss)

___
*[`pkg/util/timeutil/time_zone_util.go` line 71 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/141506#-OJ4TlFK3qRrRIlfPK2z:-OJ4bW5M9ehWX2tY-37a:b-dxcbts) ([raw file](https://github.com/cockroachdb/cockroach/blob/93449886743099e2ef2d05ebd9dcc15da7df0115/pkg/util/timeutil/time_zone_util.go#L71)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

my reasoning was that most timezone names include a `/`. for example: https://github.com/cockroachdb/cockroach/blob/ddd0076bf430e74ad79da369c163854146afa8ab/pkg/util/timeutil/lowercase_timezones_generated.go#L10-L22
</blockquote></details>

Makes sense.


<!-- Sent from Reviewable.io -->

rafiss commented:
tftr!

bors r+
